packages/block-editor/src/hooks/color.js

```diff
@@ -0,0 +1,247 @@
+import { getBlockDOMNode } from '../utils/dom';
+
+export const COLOR_SUPPORT_KEY = '__experimentalColor';
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `backgroundColor` and `textColor` attribute.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+function addAttributes( settings ) {
+	if ( ! hasBlockSupport( settings, COLOR_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	// allow blocks to specify their own attribute definition with default values if needed.
+	if ( ! has( settings.attributes, [ 'backgroundColor', 'type' ] ) ) {
+		settings.attributes = assign( settings.attributes, {
+			backgroundColor: {
+				type: 'object',
+			},
+		} );
+	}
+	if ( ! has( settings.attributes, [ 'textColor', 'type' ] ) ) {
+		settings.attributes = assign( settings.attributes, {
+			textColor: {
+				type: 'object',
+			},
+		} );
+	}
+
+	return settings;
+}
+
+/**
+ * Override props assigned to save component to inject colors classnames.
+ *
+ * @param  {Object} props      Additional props applied to save element
+ * @param  {Object} blockType  Block type
+ * @param  {Object} attributes Block attributes
+ * @return {Object}            Filtered props applied to save element
+ */
+export function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, COLOR_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	// I'd have prefered to avoid the "style" attribute usage here
+	const { backgroundColor, textColor, style } = attributes;
+
+	const backgroundClass = getColorClassName(
+		'background-color',
+		backgroundColor
+	);
+	const textClass = getColorClassName( 'color', textColor );
+	props.className = classnames( props.className, backgroundClass, textClass, {
+		'has-text-color': textColor || style?.textColor,
+		'has-background': backgroundColor || style?.backgroundColor,
+	} );
+
+	return props;
+}
+
+/**
+ * Filters registered block settings to extand the block edit wrapper
+ * to apply the desired styles and classnames properly.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+export function addEditProps( settings ) {
+	if ( ! hasBlockSupport( settings, COLOR_SUPPORT_KEY ) ) {
+		return settings;
+	}
+	const existingGetEditWrapperProps = settings.getEditWrapperProps;
+	settings.getEditWrapperProps = ( attributes ) => {
+		let props = {};
+		if ( existingGetEditWrapperProps ) {
+			props = existingGetEditWrapperProps( attributes );
+		}
+		return addSaveProps( props, settings, attributes );
+	};
+
+	return settings;
+}
+
+const ColorPanel = ( { colorSettings, clientId } ) => {
+	const { getComputedStyle, Node } = window;
+
+	const [ detectedBackgroundColor, setDetectedBackgroundColor ] = useState();
+	const [ detectedColor, setDetectedColor ] = useState();
+
+	// TODO: check performance impact of running this on each render
```

Comment:
Detected colors are updated on each render which could be bad performance-wise, but it's the only reliable way cause we  can't watch these.  So there might be some performance optimisations here. Only run the check for selected blocks...

Reply:
The variables may be inherited from a parent block e.g: --wp--background-color may have value because another parent block group set a value.
For the web CSS variables cascade automatically and we probably don't need to access the values but for mobile, a way to have the cascading mechanism may be necessary.
I guess we can have something like useGlobalStyles that returns the current style variables including the inherited values.
Maybe contrast checking can just use a useGlobalStyles hook and check the values of color and background-color and avoid the need to retrieve colors from the dom.

Reply:
I believe detecting from DOM will always be simpler and more accurate.

Reply:
> I believe detecting from DOM will always be simpler and more accurate.

I agree with that, the only thing I am not sure is if we will be able to have a reliable/performant way to use dom detection.
The background may be changed by any of the parent blocks, may this way should offer a way for parents to trigger a recomputation of colors in children.

Reply:
Is there no way to [avoid unnecessary useEffect calls](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects) in this situation?

Reply:
Is there no way to [avoid unnecessary useEffect calls](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects) in this situation?

Reply:
There's no way to know whether a useEffect is unecessary or not here :). We need to run performance tests and check the impact that's it.

Reply:
I've run performance tests between master and the paragraph refactoring branch to get meaningful results, The results are very similar between the branches. So there's no impact here.

Reply:
We have a bot that checks the bundle size, it'd be great to have another bot that runs performance comparison when there is a change in the production code :)

Reply:
💯 using the code of the bundle size one as a starting point should help here.